source/_docs/installation/synology.markdown

```diff
@@ -2,78 +2,287 @@
+You will now need to get the [spksrc framework and Docker container](https://github.com/SynoCommunity/spksrc#docker) for the compiling environment.
 
+Install Git
+#### Ubuntu/Mint
+```bash
+# sudo apt-get install git
+```
+#### Solus
 ```bash
-# ./python3 -m ensurepip
+# sudo eopkg install git
 ```
+Fork and clone spksrc, this will make a directory named “spksrc”.
+```bash
+git clone https://You@github.com/You/spksrc.git ~/spksrc
+```
+Download the spksrc Docker container:
+```bash
+docker pull synocommunity/spksrc
+```
+
+Now you need to edit 2 files.
+You need to do this to get “cross compiled module package” files, which you require for Home Assistant to able to be installed.
+These edits also enable you to use the “[Cloud](/components/cloud/)” component and fix the OpenSSL errors when using the “[Xiaomi_Aqara](/components/Xiaomi_Aqara/)” component.
 
-Use PIP to install Homeassistant package 0.64.3
 
+Edit “*~/spksrc/spksrc/spk/python3/src/requirements.txt*”, add at the end of the file:
+```
+##Cross compilation requirements for installing Home Assistant (Tested to work on 83.3).
+##In the future, the requirements may change (need newer version to work, eg), modify as needed.
+cffi==1.11.5
+bcrypt==3.1.4
+cryptography==2.3.1
+#Cross compilation requirement for installing “Warrent” module (Needed by “Cloud” Component)
+pycryptodome==3.7.2
+```
+
+Edit “*~/spksrc/spksrc/spk/python3/src/Makefile*”, add above the line that says “<b>include ../../mk/spksrc.spk.mk</b>”:
+```makefile
+# Needed to fix “_openssl.so: undefined symbol: pthread_atfork” error caused by lack of libpthread linkage on Synology (Needed for SSL and “xiaomi_aqara” component)
+export CFLAGS=-pthread
+```
+
+## {% linkable_title Compiling the Python3 package %}
+Run the container, this will make your terminal run inside the Docker container:
 ```bash
-# ./python3 -m pip install homeassistant==0.64.3
+# docker run -it -v ~/spksrc:/spksrc synocommunity/spksrc /bin/bash
+# cd spksrc/
+# make setup
+# cd spk/python3
 ```
+<p class='note'>
+You need to change the arch-XXXX to the architecture of your Synology (eg, DS115j = arch-armada370)
+For information about the Arch of your Synology look at the <a href="https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model">list of architectures on spkrc</a>
+Depending on your device, compilation may take a hour or more (significantly less if you have a SSD and a good CPU).
+</p>
 
-Create homeassistant config directory & switch to it
+Compile Python3 for your Synology model
+```bash
+# make arch-XXXX
+```
+After the compilation is done, you can find the Python package at “*~/spksrc/spksrc/packages/python3_XXXX.spk*”.
+It should be named something like “python3_armada370-6.1_3.5.5-7.spk”, of course with a possibly different arch and version.
 
+Now you need to extract your “cross compiled module package” files which you added earlier.
+For ease of use this guide will use p7zip for extracting from these archives.
```

Comment:
Why not use `tar/zip` which is likely to be in the default package set?

Reply:
For simplicity, so the user can simply copy paste a command and stuff just works.
I know no default program that can rename files inside archives without extracting.
7-Zip can and does it fast and quietly and since the guide describes installing Docker and Git, installing p7zip shouldn't be a problem.

The user also does not have to use Git and 7-Zip if he does not want/can, since the user can use the "Clone/Download" button on Github and rename all files as described inside the ".whl" files manually using any zip archiving program.

If I should change this anyway, I could write a command which unpacks, renames and repacks the file.

Reply:
I was just asking. It make sense to use the the tools which provide the easiest was to do something.

Reply:
If it makes that part better, I have made most of the commands to use usually found tools in the system.
The part for renaming the archives I have still made use of p7zip, but made it a command which will rename all files required to be renamed in all '.whl' archives found in it's current directory instead of presuming a certain file.

Reply:
Hello, I made a command which makes use of the default included commands in the system to do the renaming job, however I do feel like it is more efficient to just use the p7zip based command, as it's shorter, simply renames files inside archives without unpacking and is faster.

This one will do the same kind of big operation, but instead does the following: Do a operation on all .whl archives, unzip archive, rename files, create zip, delete tempdir, repeat.
It runs in the current directory and is basically a copy paste should just work kind of command.
```sh
rand=$RANDOM; for module in *.whl; do unzip "$module" -d "temp$rand"; find "temp$rand" -name "*x86_64-linux-gnu*" -type f | while read -r file; do mv "$file" "$(echo $file | sed "s/x86_64-linux-gnu/arm-linux-gnueabihf/")"; done; (cd "temp$rand"; echo $PWD; zip -fr0 "../$module" ./); rm -rf "temp$rand"; done
```
Should I replace the 7z based command with this?